lib/Echidna/Output/Source.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+
+module Echidna.Output.Source where
+
+import Control.Lens
+import Data.Maybe (fromMaybe, mapMaybe)
+import Data.Text (Text, unlines, pack, unpack)
+import Data.Text.Encoding (decodeUtf8)
+import Data.Text.IO (writeFile)
+import Data.List (nub)
+
+import EVM.Solidity (stripBytecodeMetadata, SourceCache, SrcMap, SolcContract, sourceLines, sourceFiles, runtimeCode, runtimeSrcmap, creationSrcmap)
+import EVM.Debug (srcMapCodePos)
+import Prelude hiding (unlines, writeFile)
+
+import qualified Data.Vector as V
+
+import qualified Data.Map as M
+import qualified Data.Set as S
+
+import Echidna.Exec 
+import Echidna.Types.Tx
+
+
+type FilePathText = Text
+
+saveCoveredCode :: Maybe FilePath -> SourceCache -> [SolcContract] -> CoverageMap -> IO ()
+saveCoveredCode (Just d) sc cs s = let fn = d ++ "/covered.txt"
+                                       cc = ppCoveredCode sc cs s 
+                                   in writeFile fn cc
+saveCoveredCode Nothing  _  _  _ = pure ()
+
+
+-- | Pretty-print the covered code
+ppCoveredCode :: SourceCache -> [SolcContract] -> CoverageMap -> Text
+ppCoveredCode sc cs s | s == mempty = "Coverage map is empty"
+                      | otherwise   =
+  let allLines = M.toList $ view sourceLines sc
+  -- ^ Collect all the possible lines from all the files
+      findFile k = fst $ M.findWithDefault ("<no source code>", mempty) k (view sourceFiles sc)
+  -- ^ Auxiliary function to get the path for each source file
+      covLines = concatMap (srcMapCov sc s) cs
+  -- ^ List of covered lines during the fuzzing campaing
+  in unlines $ map snd $ concatMap (\(f,vls) ->
+    (mempty, findFile f) :                                                   -- Add a header for each source file to show its complete path
+      filterLines covLines (map ((findFile f,) . decodeUtf8) $ V.toList vls) -- Show the source code for each file with its covered line.
+                                   ) allLines
+
+-- | Mark one particular line, from a list of lines, keeping the order of them 
+markLine :: Int -> TxResult -> FilePathText -> [(FilePathText, Text)] -> [(FilePathText, Text)] 
+markLine n r cf ls = case splitAt (n-1) ls of
+                        (xs, (f,y):ys) | f == cf -> xs ++ [(cf, markStringLine r $ unpack y)] ++ ys
+                        _                        -> map (\(f,y) -> (f, checkMarkers $ unpack y)) ls
+
+-- | Header preppend to each line
+markerHeader :: String
+markerHeader = "    |"
+
+-- | Add space for markers if necessary
+checkMarkers  :: String -> Text
+checkMarkers l@(_:_:_:_:'|':_) = pack l
+checkMarkers l                 = pack $ markerHeader ++ l
+
+-- | Add a proper marker to a line, and convert it to Text
+markStringLine  :: TxResult -> String -> Text
```

Comment:
would it be simpler to gather all markers into a set for each line and later render the line in one pass based on markers in the set?

Reply:
Yes, perhaps that will make `markStringLine` simpler, but I believe it will also make the other functions more complex. This approach keeps the types as we have them directly from the hevm functions that are used to collect the coverage information, but feel free to provide some code on how to improve it. :smile: